src/lib/operators/table_scan/column_between_table_scan_impl.cpp

```diff
@@ -54,19 +57,56 @@ void ColumnBetweenTableScanImpl::_scan_generic_segment(const BaseSegment& segmen
 
     auto typed_left_value = type_cast_variant<ColumnDataType>(_left_value);
     auto typed_right_value = type_cast_variant<ColumnDataType>(_right_value);
-    auto comparator = [typed_left_value, typed_right_value](const auto& position) {
-      return position.value() >= typed_left_value && position.value() <= typed_right_value;
-    };
 
-    _scan_with_iterators<true>(comparator, it, end, chunk_id, matches);
+    if (_left_inclusive) {
+      if (_right_inclusive) {
+        auto comparator = [typed_left_value, typed_right_value](const auto& position) {
+          return position.value() >= typed_left_value && position.value() <= typed_right_value;
+        };
+        _scan_with_iterators<true>(comparator, it, end, chunk_id, matches);
+      } else {
+        auto comparator = [typed_left_value, typed_right_value](const auto& position) {
+          return position.value() >= typed_left_value && position.value() < typed_right_value;
+        };
+        _scan_with_iterators<true>(comparator, it, end, chunk_id, matches);
+      }
+    } else {
+      if (_right_inclusive) {
+        auto comparator = [typed_left_value, typed_right_value](const auto& position) {
+          return position.value() > typed_left_value && position.value() <= typed_right_value;
+        };
+        _scan_with_iterators<true>(comparator, it, end, chunk_id, matches);
+      } else {
+        auto comparator = [typed_left_value, typed_right_value](const auto& position) {
+          return position.value() > typed_left_value && position.value() < typed_right_value;
+        };
+        _scan_with_iterators<true>(comparator, it, end, chunk_id, matches);
+      }
+    }
   });
 }
 
 void ColumnBetweenTableScanImpl::_scan_dictionary_segment(const BaseDictionarySegment& segment, const ChunkID chunk_id,
                                                           PosList& matches,
                                                           const std::shared_ptr<const PosList>& position_filter) const {
-  const auto left_value_id = segment.lower_bound(_left_value);
-  const auto right_value_id = segment.upper_bound(_right_value);
+  ValueID left_value_id;
+  if (_left_inclusive) {
+    left_value_id = segment.lower_bound(_left_value);
+  } else {
+    left_value_id = segment.upper_bound(_left_value);
+  }
+
+  ValueID right_value_id;
+  if (_right_inclusive) {
+    right_value_id = segment.upper_bound(_right_value);
+  } else {
+    right_value_id = segment.lower_bound(_right_value);
+  }
+
+  if (left_value_id == INVALID_VALUE_ID || left_value_id == right_value_id || left_value_id >= right_value_id) {
```

Comment:
the second condition is unnecessary, sure. but what do you mean with moving the third condition up in the code? I just get the values in the lines above. I could move it into an assert, but because it is possible and no error case, this does not seem useful.

Reply:
Dropping the second condition should be fine for us